pallets/interest-accrual/src/lib.rs

```diff
@@ -132,82 +160,70 @@ pub mod pallet {
 		DebtAdjustmentFailed,
 		/// Emits when the interest rate was not used
 		NoSuchRate,
+		/// Emits when a historic rate was asked for from the future
+		NotInPast,
 	}
 
-	// TODO: add permissionless extrinsic to update any rate
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_initialize(_: T::BlockNumber) -> Weight {
+			let mut count = 0;
+			let then = LastUpdated::<T>::get();
+			let now = Self::now();
+			LastUpdated::<T>::set(Self::now());
+			let delta = Self::now() - then;
+			let bits = Moment::BITS - delta.leading_zeros();
+			Rate::<T>::translate(|per_sec, mut rate: RateDetailsOf<T>| {
+				count += 1;
+				Self::calculate_accumulated_rate(per_sec, rate.accumulated_rate, then, now)
+					.ok()
+					.map(|new_rate| {
+						rate.accumulated_rate = new_rate;
+						rate
+					})
+			});
+			T::DbWeight::get().reads_writes(2, 1)
+				+ count
+					* (T::DbWeight::get().reads_writes(1, 1)
+						+ T::Weights::calculate_accumulated_rate(bits))
+		}
+	}
 
 	impl<T: Config> Pallet<T> {
 		pub fn get_current_debt(
 			interest_rate_per_sec: T::InterestRate,
 			normalized_debt: T::Balance,
 		) -> Result<T::Balance, DispatchError> {
-			Rate::<T>::try_mutate(
-				interest_rate_per_sec,
-				|rate_details| -> Result<T::Balance, DispatchError> {
-					let rate = if let Some(rate) = rate_details {
-						let new_accumulated_rate = Self::calculate_accumulated_rate(
-							interest_rate_per_sec,
-							rate.accumulated_rate,
-							rate.last_updated,
-						)
-						.map_err(|_| Error::<T>::DebtCalculationFailed)?;
-
-						rate.accumulated_rate = new_accumulated_rate;
-						rate.last_updated = Self::now();
-
-						rate
-					} else {
-						*rate_details = Some(RateDetails {
-							accumulated_rate: One::one(),
-							last_updated: Self::now(),
-						});
-						rate_details.as_mut().expect("RateDetails now Some. qed.")
-					};
-
-					let debt = Self::calculate_debt(normalized_debt, rate.accumulated_rate)
-						.ok_or(Error::<T>::DebtCalculationFailed)?;
-					Ok(debt)
-				},
-			)
+			if let Some(rate) = Rate::<T>::get(interest_rate_per_sec) {
+				Self::calculate_debt(normalized_debt, rate.accumulated_rate)
+					.ok_or(Error::<T>::DebtCalculationFailed.into())
+			} else {
+				Err(Error::<T>::NoSuchRate.into())
```

Comment:
This if-let-else could probably just be an `ok_or`. Not sure why I wrote it this way